packages/components/src/public/ExtendedMap.ts

```diff
@@ -0,0 +1,145 @@
+type KeyType = string | number | symbol;
+type MapType<K extends KeyType, V> = Record<K, V> | Map<K, V>;
+type Mapper<K extends KeyType, V, T> = (value: V, key: K, original: ExtendedMap<K, V>) => T;
+type ArrayMapper<V, T> = (value: V, index: number, array: V[]) => T;
+type FilterFn<K extends KeyType, V> = (value: V, key: K, original: ExtendedMap<K, V>) => boolean;
+type Reducer<K extends KeyType, V, T> = (result: T, value: V, key: K, original: ExtendedMap<K, V>) => T;
+// TODO: convert the QueryInfo columns and views to use BetterMap, then find all the areas we are mapping or iterating
+//  over their values and replace them with usages of map, reduce, etc. from here.
+//  Most notably insertColumns should use mergeAt.
+/**
+ * ExtendedMap is an extended version of the built in Map class. It has an improved constructor (that takes Records,
+ * Map, or ExtendedMap objects), as well as several convenience methods for mapping, reducing, and filtering the map or
+ * values. This class is an Ordered Map, because it extends the Map class which is ordered.
+ */
+export class ExtendedMap<K extends KeyType, V> extends Map {
+    constructor(...data: Array<Record<KeyType, V> | Map<KeyType, V>>) {
```

Comment:
Should the constructor take immutable map as well?

Reply:
I don't think it should. You can always convert an immutable map to ExtendedMap via `new ExtendedMap(myObj.toJS());`.